### PR TITLE
[master] Transport dependencies reform

### PR DESCRIPTION
**Pull request by art-gor**
## Description

Removed MHD, jsonrpccpp backends, they all spinlocked and we were unable to detect performance bottlenecks. Now we can
.

## Backward Compatibility

- [x] This is not a breaking change
- [ ] This is a breaking change

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [x] small-scale cloud test

### CODE REVIEW

**Comment by valeryz** [comment]
Haven't yet looked in detail at the change itself, but it looks like it should be two separate PRs: 

1) vendoring of jsonrpccpp, (@yaron-zilliqa could you please have a look of this is the best option, and if there's no better option with vcpkg), 
2) removal of MHD, and other reform.

**Comment by CSideSteve** [comment]
Built and tested the code on Ubuntu 22.10 and Apple Mac M2 using clnag 14 C++ (LLVM).
    Both work and perform as expected.

Debugged through the EvmClient code to inspect code path and happy that old modifications can be discarded and the new implementation does not exhibit previous bugs exhibited by jsonrpccpp socket layer.

Number of files in branch seems overly large, I suspect this will reduce substantially when depends files are moved into platform vcpkg or successor.

Overall very pleased with the way this branch performs, need to eliminate some compiler warning on very new compilers, but probably not in this branches code as such.

Would be interested to see cpu profiles on staging and testnet, I can only imagine they are sunstantially flatter  than the the code it is going to supercede.

File: src/libServer/APIServer.h [hunk 1]
```diff
@@ -50,6 +54,9 @@ class APIServer {
     /// Limit in bytes for POST bodies of incoming requests (security)
     size_t inputBodyLimit = 5 * 1024 * 1024;
```
Comment (yaron-zilliqa) [inline src/libServer/APIServer.h:55]: A suggestion: use the unit as part of the name (more readable and saves the developer the need to look up the comments): _limitBodyLimitBytes_

File: src/libServer/GetWorkServer.cpp [hunk 2]
```diff
@@ -34,10 +34,22 @@ using namespace jsonrpc;
 
 static ethash_mining_result_t FAIL_RESULT = {"", "", 0, false};
 
+namespace {
+
+std::shared_ptr<rpc::APIServer> GetServerConnector() {
+  rpc::APIServer::Options options;
+  options.threadPoolName = "GetWork";
+  options.numThreads = 2;
```
Comment (yaron-zilliqa) [inline src/libServer/GetWorkServer.cpp:42]: Any specific reason why it's 2? comment?

File: src/libServer/LocalAPIServer.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  m_thread.emplace([this] { WorkerThread(); });
+
+  return true;
+}
+
+bool LocalAPIServer::StopListening() { return DoStop(); }
+
+namespace {
+
+void WakeAcceptor(const boost::asio::ip::tcp::endpoint& endpoint) {
```
Comment (yaron-zilliqa) [inline src/libServer/LocalAPIServer.cpp:84]: What's the point of this?

File: src/libServer/LocalAPIServer.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    std::string readBuffer;
+    std::string response;
+
+    while (m_started) {
+      boost::asio::ip::tcp::socket socket(m_asio);
+
+      boost::system::error_code ec;
+      m_acceptor->accept(socket, ec);
+      if (ec && m_started) {
+        throw std::runtime_error(ec.message());
```
Comment (yaron-zilliqa) [inline src/libServer/LocalAPIServer.cpp:120]: Consider throwing a system_error constructed from ec.

File: src/libServer/UnixDomainSocketClient.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    // Write the JsonRpc
+    try {
+      boost::asio::write(unixDomainSocket,
+                         boost::asio::buffer(toSend, toSend.length()));
+    } catch (std::exception& e) {
+      if (LOG_SC) {
+        LOG_GENERAL(INFO, "Exception calling write " << e.what());
+      }
+      throw e;
```
Comment (yaron-zilliqa) [inline src/libServer/UnixDomainSocketClient.cpp:55]: Can just be: throw; to rethrow the same exception.

File: src/libServer/UnixDomainSocketClient.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+#include <jsonrpccpp/client.h>
+
+// This is a Custom socket handler using asio
+// for the client connection to evm-ds and scilla server
+
+namespace rpc {
+
+class UnixDomainSocketClient : public jsonrpc::IClientConnector {
+ public:
+  UnixDomainSocketClient(const std::string& path) : m_path(path){};
```
Comment (yaron-zilliqa) [inline src/libServer/UnixDomainSocketClient.h:30]: explicit?

File: tests/Server/Test_ScillaIPCServer.cpp [hunk 7]
```diff
@@ -68,10 +69,10 @@ BOOST_AUTO_TEST_CASE(test_connection) {
 
 // Simple non-map query.
 BOOST_AUTO_TEST_CASE(test_query_simple) {
-  UnixDomainSocketServer s(SCILLA_IPC_SOCKET_PATH);
+  rpc::UnixDomainSocketServer s(SCILLA_IPC_SOCKET_PATH);
```
Comment (yaron-zilliqa) [inline tests/Server/Test_ScillaIPCServer.cpp:72]: Perhaps move this to some fixtures class so it doesn't have to be repeated for each test?

File: src/libData/AccountData/ScillaClient.cpp [hunk 8]
```diff
@@ -137,10 +138,9 @@ bool ScillaClient::CheckClient(uint32_t version, bool enforce) {
     return false;
   }
 
-  std::shared_ptr<jsonrpc::UnixDomainSocketClient> conn =
-      std::make_shared<jsonrpc::UnixDomainSocketClient>(
-          SCILLA_SERVER_SOCKET_PATH +
-          (ENABLE_SCILLA_MULTI_VERSION ? ("." + std::to_string(version)) : ""));
+  auto conn = std::make_shared<rpc::UnixDomainSocketClient>(
```
[the remark this answers is not in the document]
Comment (yaron-zilliqa) [reply src/libData/AccountData/ScillaClient.cpp:141]: That point is that std::unique_ptr is "lighter" than std::shared_ptr, so I would default to using this unless necessary.

File: src/libServer/LocalAPIServer.cpp (hunk 3 above)
Comment (yaron-zilliqa) [reply src/libServer/LocalAPIServer.cpp:84]: It seems to be called from DoStop() so I'm not sure my guess would be correct.

**Review by CSideSteve — DISMISSED** [review]
As long as platform approve or change the packaging of jsonrpc I am happy to approve this version after it has run on evmdev. the src/depends directory is on its way out, would be sad to add to it.

As we will have to support jsonrpc going forward it would be beneficial to remove our dependency on it completely and use either the new rpccpp from same author or a suitable alternative as this will reduce code size and modernise source in a few areas.

**Comment by valeryz** [comment]
Ack, @art-gor please don't submit without approval from Platform regarding the jsonrpc package vs. vendoring in 'depends'.

File: src/libServer/GetWorkServer.cpp (hunk 2 above)
Comment (valeryz) [reply src/libServer/GetWorkServer.cpp:42]: I believe the question here was to just put a comment in the code, not to make it a magic constant, @art-gor

**Review by yaron-zilliqa — APPROVED** [review]

**Pull request merged**